kernel/src/process_load.rs

```diff
@@ -0,0 +1,791 @@
+                    self.footer.set(0);
+                }
+                FooterCheckResult::NoProcess | FooterCheckResult::BadFooter => {
+                    // Go to next process
+                    self.process.set(self.process.get() + 1);
+                    self.footer.set(0)
+                }
+                FooterCheckResult::FooterNotCheckable => {
+                    // Go to next footer
+                    self.footer.set(self.footer.get() + 1);
+                }
+                FooterCheckResult::Error => {
+                    return Err(ProcessLoadError::InternalError);
+                }
+            }
+        }
+    }
+}
+
+// Returns whether a footer is being checked or not, and if not, why.
+// Iterates through the footer list until if finds `next_footer` or
+// it reached the end of the footer region.
+fn check_footer(
+    popt: Option<&'static dyn Process>,
+    verifier: &'static dyn AppVerifier,
+    next_footer: usize,
+) -> FooterCheckResult {
+    popt.map_or(FooterCheckResult::NoProcess, |process| {
+        if config::CONFIG.debug_process_credentials {
+            debug!(
+                "Checking: Checking {} footer {}",
+                process.get_process_name(),
+                next_footer
+            );
+        }
+        let footers_position_ptr = process.get_addresses().flash_integrity_end;
+        let mut footers_position = footers_position_ptr as usize;
+
+        let flash_start_ptr = process.get_addresses().flash_start as *const u8;
+        let flash_start = flash_start_ptr as usize;
+        let flash_integrity_len = footers_position - flash_start;
+        let flash_end = process.get_addresses().flash_end as usize;
+        let footers_len = flash_end - footers_position;
+
+        if config::CONFIG.debug_process_credentials {
+            debug!(
+                "Checking: Integrity region is {:x}-{:x}; footers at {:x}-{:x}",
+                flash_start,
+                flash_start + flash_integrity_len,
+                footers_position,
+                flash_end
+            );
+        }
+        let mut current_footer = 0;
+        let mut footer_slice = unsafe { slice::from_raw_parts(footers_position_ptr, footers_len) };
+        let binary_slice = unsafe { slice::from_raw_parts(flash_start_ptr, flash_integrity_len) };
+        while current_footer <= next_footer && footers_position < flash_end {
+            let parse_result = tock_tbf::parse::parse_tbf_footer(footer_slice);
+            match parse_result {
+                Err(TbfParseError::NotEnoughFlash) => {
+                    return FooterCheckResult::PastLastFooter;
+                }
+                Err(TbfParseError::BadTlvEntry(t)) => {
+                    if config::CONFIG.debug_process_credentials {
+                        debug!("Checking: Bad TLV entry, type: {:?}", t);
+                    }
+                    return FooterCheckResult::BadFooter;
+                }
```

Comment:
This case is a bit tricky. If the footer region does not extend to the end of the entire app flash region, then this loop will attempt to parse a TLV past the end of the footers, which will almost certainly be invalid. That triggers this bad TLV case. Since this returns BadFooter, that essentially causes the credential checking to stop/be invalid for this app.

What is the intended behavior here? We need to allow padding past the footer for the case where the app needs fewer than 8 bytes of padding to get to its desired size.

Reply:
I think that if there isn't enough space for a footer, `parse_tbf_footer` returns `TbfParseError::NotEnoughFlash`, which results in a `FooterCheckResult::PastLastFooter`, which will trigger the end of the checking state machine. This is worth testing, though.

As for empty space after the last footer, this is tricky. If it can be arbitrary bits, then this means it might be parsed as footers. Should we specify that padding needs to be 0s, and reserve 0 as Padding?

Reply:
> litex_sim with `debug_load_processes`:
> 
> ```
> Verilated LiteX+VexRiscv: initialization complete, entering main loop.
> Loading processes from flash=0x00080000-0x000FFFFF into sram=0x40003400-0x4FFFFFFF
> Loading process from flash=0x00080000-0x000FFFFF into sram=0x40003400-0x4FFFFFFF
> No more processes to load: App flash does not match requested address. Actual:0x80020, Expected:0x80060.
> Checking: no checking, load and run all processes
> ```

Found it -- it was how the offset was being calculated; for fixed apps I had forgotten to revert to the old protected region calculation. This now passes CI on Litex and I verified it also loads multiple dynamic processes on imix.

Reply:
I checked that a process which has some true padding (space after its credentials that is too small for another credential) compile, elf2tab, and load properly. Furthermore, processes after such a process load properly. I did this my putting a static initializer in c_hello and scaling the size of the array until it created bumped to the next power-of-2. Then I reduced it by 8 -- 4 to bring it to the exact right size for a power of 2, then another 4 to make it smaller.

Reply:
Good point. I ran into this because I added a `tockloader tbf-delete-credential` feature to tockloader, and it just naively removes the specified credential TLV and doesn't currently change the reserved space, so when the app was flashed tockloader added empty bytes of padding.




I'm still thinking about this, but I'm kind of leaning towards specifying that if an app has a footer, then there can be at most 7 bytes of "empty space" between the end of the footer and the `total_size` of the app. Then, even if there is padding, the parsing will fail with `NotEnoughFlash` which is OK. If the app has more than 7 bytes of padding that is a poorly constructed app and failing with a bad footer (or something) is totally fine.

Reply:
I assume this means that if you have footers (`total_length` - `binary_end` >= 8), then footers must fill the region? There's still the main header.